### ISSUE TITLE
[Proposal] Adopt Snake Case for Variable Naming Conventions

### ISSUE THREAD

**Issue by irrld** [GlistEngine/GlistEngine#949]
I propose a change in our variable naming conventions to improve readability and maintainability in GlistEngine. Specifically, I suggest the following:

1. **Local Variables (variables defined inside functions):** Use `snake_case` for local variable names (this includes function parameters).
2. **Member Variables (variables defined inside classes):** Use `snake_case_` (snake case with a trailing underscore) for member variable names.
3. **Static Variables:** Use `snake_case_` (snake case with a trailing underscore), same as member variables.
4. **Static Const:** Unchanged, UPPER_CASE is used.
5. **Enums:** Unchanged, UPPER_CASE is used with the prefix of the enum name. If the enum is an `enum class`, the prefix can be omitted as it requires the type to be written (e.g. `TabPosition::LEFT`).
6. **Classes:** Unchanged, `PascalCase` is used, with a `g` prefix indicating the class is from GlistEngine.
7. **Functions:** Unchanged, `camelCase` is used.
8. **Macros:** Unchanged, UPPER_CASE is used.
9. **Files:** Unchanged, `PascalCase` is used, with an optional `g` prefix indicating the class is from GlistEngine. Should be the same as the class name defined inside.
10. **Usage of `auto`:** Should be avoided unless it makes the code **clearer** to read, should not be used to avoid typing a name. For example, it can be used for iterators as it requires a long type definition.

**Rationale:**

- **Readability:** `snake_case` is a widely accepted convention for variable names in many programming communities, making our code more readable and accessible to contributors.
- **Clear Distinction:** Using different naming conventions helps in clearly distinguishing between different types of identifiers. For instance, `helloWorld` could be a variable or a function, whereas `hello_world_` is clearly a member variable and `hello_world` is a local variable.
- **Differentiation:** The trailing underscore for member variables will clearly differentiate them from local variables, eliminating the need for `this->` usage.

**Implementation:**

This change would involve:
- Adopting `snake_case` and `snake_case_` conventions for all newly written code.
- Existing code will remain unchanged unless it is being refactored, in which case the new conventions should be applied.

This change might seem unnecessary, but how we name things is important. We already spend a lot of time trying to find bugs, and spending more to understand what a variable name stands for is unnecessary work for the developers. I personally think these changes makes working on the engine much more pleasant. 

Please feel free to share your thoughts.

**Comment by sorbatdev** [comment]
I agree on this approach, with an exception and some additions:

- I believe that member variables should be even more obvious to the developer. Therefore in my opinion they should start with the `m_` prefix, and continue with the snake_case.
- Arguments of functions should also have a prefix, `t_` and continue with a snake_case.
- Template arguments should be addressed and they should be PascalCase.

**Comment by irrld** [comment]
Edited and added template arguments.

**Comment by javertus** [comment]
Well, in my opinnion this is better when comapred to our current namings. Also using `PascalCase` for Function names would be better

**Comment by UsifDev** [comment]
I agree with this, I'm specifically not a fan of naming member variables with lowercase.
I think it's better to indicate terms of a long variable in a clearer form. I'm not sure if snake_case is the best for this but it's definitely better than lowercase.

Code will become more complex over time and naming complex things that need to express a lot of detail shouldn't be a problem and its solution shouldn't be to compress it into an acronym or shortened version too (unless it's well known or special for that within the class context, like arima model or temp for temporary (may be confused with temperature when working on a fluid simulation))

In general, I believe the simplest solution is to use spacing, order, size and/or groupings to avoid accidentally limiting the amount of information that the developer wants to provide.
 
It may seem like too much unnecessary detail, but the effect is very visible when working on large files in contrast to the existing large files especially in the .cpp files and it may increase productivity by spending less time debugging or losing focus on a problem-solving task because your mind shifted all of its attention to find one thing.

For example: gridxytintfact can be a variable to partially control the influence of the grid's axes on the color of points within a limited distance away from each axis. it would be more readable to just write it as points_color_grid_xy_axes_tint_factor (points_color to express that this applies to the color of all points and may be omitted)

_side note: writing this made me realize I have errors in the gGrid class files_ 🙃. _I copied the old axes features from gRenderer assuming it's for individual axis because of the usage of the singular form and now realized that the intent was for both axes of a grid to have the same gridxycolor or gridxzcolor etc. This could've been expressed in the code like this grid_xy_axescolor. Regardless, if a grid controls the color of an axis, then the other grid sharing that axis can't use a different color. This logic is confusing and unnecessary. Each axis color should be controlled alone_

### To strengthen the proposal even more, here's a better example from the gGUIManager header file:
![image](https://github.com/user-attachments/assets/5f1e3f88-fb2c-4f2c-8670-f4c95f96a861)
![image](https://github.com/user-attachments/assets/64967125-183e-43f3-bcc3-c316d7e73370)

It's hard to read a variable and then go do something then go back to the one you focused on moments ago because of too many terms bunched up together so you have to re-search for it again.

If snake_case was used, I think the states/adjectives of relevant terms can just be prefix-ed without underscore to make the most different/important terms stand out like theme_pressedbutton_fontcolor instead of theme_pressed_button_font_color

The spacing created by the underscore within the grouping combined with term ordering, adds sort of a visual structure, making it easy to go back to the thing you were looking at moments ago, which is very helpful for quick debugging especially if it's an error in someone else's code.

_try finding the variables related with "font" in both images, did you notice that font is always behind color, which is always the last term, and that it doesn't appear with the term "background" at all? that's the effect I'm referring to_

**Comment by AlperTurhann** [comment]
The suggested use of snake_case for local variables and snake_case_ for member variables is particularly appealing. This clear distinction will make it much easier to understand the scope and purpose of variables at a glance, potentially reducing bugs and speeding up code reviews.

While there might be some short-term challenges in adapting to these new conventions, I believe the long-term benefits in code clarity and maintainability far outweigh any temporary inconvenience.

Moreover, aligning our naming conventions more closely with widely accepted practices in the broader C++ community will make our codebase more accessible to new contributors and more consistent with what many developers expect.

**Comment by fdesc** [comment]
This is a good idea but i have another idea to add on top (this might be a little bit overkill) what if we made a small program or kind of a script that would substitute the naming depending on the developer's desires.

**Comment by irrld** [comment]
Added define guard, forward declarations and inheritance rules. These are the same rules currently followed by the engine.

**Comment by Kamianz** [comment]
I prefer snake_case over flatcase for a few reasons. Snake_case makes long variable names easier to read and understand. For instance, current_player_health_status is much clearer than currentplayerhealthstatus, where repeated letters can make it harder to read. This improved readability makes the code more maintainable and easier to work with in the long run.

**Comment by kalimbatrr** [comment]
I agree with this snake_case campaign and hoping to make easier for both junior and senior labelled developers

**Comment by sfidan42** [comment]
I agree with this proposal 👍. The separation between local variables (snake_case) and member variables (snake_case_) will make the codebase much clearer and easier to navigate. The consistency across identifiers also lowers the barrier for new contributors. Thanks for putting this together!

**2 Improvements:**
1. Use of k_UPPER_CASE for static const
2. Use of m_snake_case for member variables
